clients/da-vinci-client/src/main/java/com/linkedin/davinci/blobtransfer/NettyP2PBlobTransferManager.java

```diff
@@ -51,19 +53,42 @@ public CompletionStage<InputStream> get(String storeName, int version, int parti
     if (discoverPeers == null || discoverPeers.isEmpty()) {
       throw new VenicePeersNotFoundException("No peers found for the requested blob");
     }
+    LOGGER
+        .info("Discovered peers {} for store {} version {} partition {}", discoverPeers, storeName, version, partition);
     for (String peer: discoverPeers) {
-      try {
-        // TODO: add some retry logic or strategy to choose the peers differently in case of failure
-        // instanceName comes as a format of <hostName>_<applicationPort>
-        String chosenHost = peer.split("_")[0];
-        LOGGER.info("Chosen host: {}", chosenHost);
-        inputStream = nettyClient.get(chosenHost, storeName, version, partition);
-        return inputStream;
-      } catch (Exception e) {
-        LOGGER.warn("Failed to connect to peer: {}", peer, e);
+      int retryCount = 0;
+      while (retryCount < MAX_RETRIES_FOR_BLOB_TRANSFER_PER_HOST) {
+        try {
+          // instanceName comes as a format of <hostName>_<applicationPort>
+          String chosenHost = peer.split("_")[0];
+          LOGGER.info("Attempt {} to connect to host: {}", retryCount + 1, chosenHost);
+          CompletionStage<InputStream> inputStreamStage = nettyClient.get(chosenHost, storeName, version, partition);
+          CompletableFuture<InputStream> inputStreamFuture = inputStreamStage.toCompletableFuture();
+
+          InputStream inputStream = inputStreamFuture.get(30, TimeUnit.MINUTES);
```

Comment:
If we fail 3 times, doesn't it mean potentially we has wait up to 90 mins and makes no progress and we are doing this for every peer?? That would be unacceptable for use cases that needs fast bootstrap. Ideally, we should only try to connect and retrieve blobs up to one (or two) peer(s) before giving up the blob transfer and using kafka ingestion.
How about we split the `nettyClient.get` into two parts? The first part is for connecting and the second is to send the GET request to request blobs?

So we can have something like
```
CompletableFuture <Channel> connectionFuture = nettyClient.connect(chosenHost);
// if it connects, then
   CompletableFuture <InputStream> fileTransferFuture = nettyClient.get(storeName, version, partition); // attach handlers and send the GET request

// if it fails, then try next peer until a valid peer is found.
```
What do you think?

Reply:
Yes, it's possible that after 90 minutes the transfer isn't completed and we have to move on to the next hosts. However, this would be a corner case where every host has a good connection, but the file transfer isn’t finished within 30 minutes. 

As for giving up after a few hosts, I have some doubts. In perf tests, there are 10+ hosts in the discoverPeers list, but we only need one specific host which has snapshot. Therefore, deciding to abandon the process based on the number of hosts may not be appropriate. Instead, let's establish a target timeframe. For instance, if the blob transfer isn't completed in 5 hours, we should give up and proceed to Kafka ingestion.

It may not be necessary to separate `nettyClient.get`, as we want to be able to retry or skip both connection issues and file transfer problems.